core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/ApacheSender43.java

```diff
@@ -39,22 +44,30 @@
  */
 final class ApacheSender43 implements ApacheSender {
 
-    private final CloseableHttpClient httpClient;
-
+    private final AtomicReference<CloseableHttpClient> httpClientRef = new AtomicReference<>();
+    private volatile boolean isClientInitialized = false;
+    private final ExecutorService initializer = Executors.newSingleThreadExecutor();
     public ApacheSender43() {
-        PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
-        cm.setMaxTotal(DEFAULT_MAX_TOTAL_CONNECTIONS);
-        cm.setDefaultMaxPerRoute(DEFAULT_MAX_CONNECTIONS_PER_ROUTE);
+        initializer.execute(new Runnable() {
+
+            @Override
+            public void run() {
+                PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
+                cm.setMaxTotal(DEFAULT_MAX_TOTAL_CONNECTIONS);
+                cm.setDefaultMaxPerRoute(DEFAULT_MAX_CONNECTIONS_PER_ROUTE);
 
-        httpClient = HttpClients.custom()
-                .setConnectionManager(cm)
-                .useSystemProperties()
-                .build();
-         }
+                httpClientRef.compareAndSet(null, HttpClients.custom()
```

Comment:
This sets the httpClient atomically. If two initialization procedures are running at the same time (somehow), the compareAndSet call will let first writer win.

Reply:
I see so you are using a SingleThreadExecutor to ensure that the initialization runs on one and only one thread and then using AtomicReference to ensure that this gets setted atomically . AtomicReference is nice when you want to share the state of the same immutable object and might want to change that shared object's value from one thread to another. This ensures atomicity and is slightly better in terms of performance. But here I am not sure you are really sharing between threads because you are using SingleThreadExecutor which I believe is a factory method and essentially runs on Single Thread. Even classical locking with Volatile, should also work here in my understanding. I can be wrong though, happy to understand your thought process.

https://stackoverflow.com/questions/3964211/when-to-use-atomicreference-in-java

Reply:
The AtomicReference is to guarantee the reads/writes to this reference are guarded without the need for another lock.

Reply:
Yup thanks for clarification. Makes sense. Should we use as a global practice throughout SDK?